crates/compiler/mono/src/ir/pattern.rs

```diff
@@ -0,0 +1,1815 @@
+                    IntLiteral(_, _)
+                    | FloatLiteral(_, _)
+                    | DecimalLiteral(_)
+                    | BitLiteral { .. }
+                    | EnumLiteral { .. }
+                    | StrLiteral(_)
+                    | Voided { .. } => None,
+                }
+            }
+            Stack(stack) => {
+                while let Some((pat, layout)) = stack.pop() {
+                    match pat {
+                        Pat(pattern) => match pattern {
+                            Identifier(symbol) => return (*symbol, layout).into(),
+                            As(pat, symbol) => {
+                                stack.push((Pat(pat), layout));
+                                return (*symbol, layout).into();
+                            }
+                            RecordDestructure(destructs, _) => stack.extend(
+                                destructs
+                                    .iter()
+                                    .map(|destruct| {
+                                        (RecordDestruct(&destruct.typ), destruct.layout)
+                                    })
+                                    .rev(),
+                            ),
+                            TupleDestructure(destructs, _) => stack.extend(
+                                destructs
+                                    .iter()
+                                    .map(|destruct| (Pat(&destruct.pat), destruct.layout))
+                                    .rev(),
+                            ),
+                            NewtypeDestructure { arguments, .. } | AppliedTag { arguments, .. } => {
+                                stack.extend(arguments.iter().map(|(p, l)| (Pat(p), *l)).rev())
+                            }
+                            OpaqueUnwrap { argument, .. } => {
+                                stack.push((Pat(&argument.0), layout));
+                            }
+                            List {
+                                element_layout,
+                                elements,
+                                ..
+                            } => {
+                                stack.extend(
+                                    elements.iter().map(|p| (Pat(p), *element_layout)).rev(),
+                                );
+                            }
+                            IntLiteral(_, _)
+                            | FloatLiteral(_, _)
+                            | DecimalLiteral(_)
+                            | BitLiteral { .. }
+                            | EnumLiteral { .. }
+                            | Underscore
+                            | StrLiteral(_)
+                            | Voided { .. } => {}
+                        },
+                        PatternBindingWork::RecordDestruct(_) => todo!(),
+                    }
+                }
+
+                *self = Done;
+                None
+            }
+        }
+    }
+}
```

Comment:
@folkertdev this is the new iterator. I went ahead and moved a bunch of the pattern stuff to a new file.